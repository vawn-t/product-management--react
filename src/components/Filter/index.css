.filter-container {
  display: flex;
  width: 100%;
  justify-content: center;
  flex-direction: column;
  margin-top: 2rem;
}
