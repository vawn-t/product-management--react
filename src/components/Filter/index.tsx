// Library
import React, { memo } from 'react';

// Constants
import { FilterOrderOptions, ProductTypes } from '@constants/index';

// Components
import { SelectItem, Button } from '@components/index';

// Styles
import './index.css';

export type FilterProps = {
  typeFilterOptions: ProductTypes[];
  priceFilterOptions: FilterOrderOptions[];
  currentFilterTypeParam?: ProductTypes;
  currentFilterPriceParam?: FilterOrderOptions;
  onTypeChange: (value: string) => void;
  onPriceChange: (value: string) => void;
  onClearFilters: () => void;
};

const Filter: React.FC<FilterProps> = ({
  typeFilterOptions,
  priceFilterOptions,
  currentFilterTypeParam = '',
  currentFilterPriceParam = '',
  onTypeChange,
  onPriceChange,
  onClearFilters
}) => (
  <div className='filter-container'>
    <div className='filter-item'>
      <SelectItem
        label='Type'
        options={typeFilterOptions}
        name='type-option'
        onSelectChange={onTypeChange}
        defaultValue={currentFilterTypeParam}
      />
    </div>
    <div className='filter-item'>
      <SelectItem
        label='Price'
        options={priceFilterOptions}
        name='price-filter'
        onSelectChange={onPriceChange}
        defaultValue={currentFilterPriceParam}
      />
    </div>
    <Button onClick={onClearFilters}>Clear filter</Button>
  </div>
);

export default memo(Filter);
