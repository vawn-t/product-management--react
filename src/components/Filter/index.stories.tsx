// Libraries
import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

// Components
import Filter from './index';

// Styles
import '@assets/styles/reset.css';
import '@assets/styles/App.css';
import '@assets/styles/variables.css';

// Constants
import { PRODUCT_TYPE_LIST, ORDER_OPTIONS } from '@constants/index';

export default {
  title: 'Components/Filter',
  component: Filter
} as ComponentMeta<typeof Filter>;

const Template: ComponentStory<typeof Filter> = (args) => <Filter {...args} />;

export const Default = Template.bind({});
Default.args = {
  typeFilterOptions: PRODUCT_TYPE_LIST,
  priceFilterOptions: ORDER_OPTIONS
};
