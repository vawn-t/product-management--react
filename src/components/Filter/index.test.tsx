// Libraries
import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { screen } from '@testing-library/dom';
import { create } from 'react-test-renderer';
import '@testing-library/jest-dom';

// Components
import Filter, { FilterProps } from './index';

// Constants
import {
  FilterOrderOptions,
  ORDER_OPTIONS,
  ProductTypes,
  PRODUCT_TYPE_LIST
} from '@constants/types';

describe('Filter component', () => {
  const defaultProps: FilterProps = {
    typeFilterOptions: PRODUCT_TYPE_LIST,
    priceFilterOptions: ORDER_OPTIONS,
    onTypeChange: jest.fn(),
    onPriceChange: jest.fn(),
    onClearFilters: jest.fn()
  };

  let container: HTMLElement;
  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    container.remove();
    jest.clearAllMocks();
    cleanup();
  });

  test('should render correctly', () => {
    const defaultTree = create(<Filter {...defaultProps} />).toJSON();
    expect(defaultTree).toMatchSnapshot();
  });

  test('should render correct params', () => {
    const { getByText } = render(
      <Filter
        {...defaultProps}
        currentFilterPriceParam={FilterOrderOptions.Desc}
        currentFilterTypeParam={ProductTypes.Tablet}
      />
    );

    expect(getByText(FilterOrderOptions.Desc)).toBeInTheDocument();
    expect(getByText(ProductTypes.Tablet)).toBeInTheDocument();
  });

  test('should onClearFilters be called', () => {
    const clearMock = jest.fn();

    render(<Filter {...defaultProps} onClearFilters={clearMock} />);

    const clearBtn: HTMLButtonElement = screen.getByRole('button');

    fireEvent.click(clearBtn);

    expect(clearMock).toHaveBeenCalled();
  });

  test('should onTypeChange is called', () => {
    const changeTypeMock = jest.fn();

    render(<Filter {...defaultProps} onTypeChange={changeTypeMock} />, {
      container
    });

    const select: HTMLElement | null = container.querySelector(
      'select[name="type-option"]'
    );

    if (select) {
      expect(select).toBeTruthy();

      fireEvent.change(select);

      expect(changeTypeMock).toHaveBeenCalled();
    }
  });

  test('should onPriceChange is called', () => {
    const changePriceMock = jest.fn();

    render(<Filter {...defaultProps} onPriceChange={changePriceMock} />, {
      container
    });

    const select: HTMLElement | null = container.querySelector(
      'select[name="price-filter"]'
    );

    if (select) {
      fireEvent.change(select);

      expect(changePriceMock).toHaveBeenCalled();
    }
  });
});
