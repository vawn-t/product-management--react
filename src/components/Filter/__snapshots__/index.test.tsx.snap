// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter component should render correctly 1`] = `
<div
  className="filter-container"
>
  <div
    className="filter-item"
  >
    <div
      className="select-wrapper"
    >
      <label
        htmlFor="type-option"
      >
        Type
        :
      </label>
      <select
        className="select"
        data-testid="select-option"
        disabled={false}
        id="type-option"
        name="type-option"
        onChange={[Function]}
        value=""
      >
        <option
          value=""
        >
          Select
        </option>
        <option
          value="Phone"
        >
          Phone
        </option>
        <option
          value="Tablet"
        >
          Tablet
        </option>
        <option
          value="Laptop"
        >
          Laptop
        </option>
      </select>
    </div>
  </div>
  <div
    className="filter-item"
  >
    <div
      className="select-wrapper"
    >
      <label
        htmlFor="price-filter"
      >
        Price
        :
      </label>
      <select
        className="select"
        data-testid="select-option"
        disabled={false}
        id="price-filter"
        name="price-filter"
        onChange={[Function]}
        value=""
      >
        <option
          value=""
        >
          Select
        </option>
        <option
          value="asc"
        >
          asc
        </option>
        <option
          value="desc"
        >
          desc
        </option>
      </select>
    </div>
  </div>
  <button
    className="btn btn-default"
    disabled={false}
    onClick={[MockFunction]}
  >
    Clear filter
  </button>
</div>
`;
